crypto/storage/vault.go

```diff
@@ -0,0 +1,118 @@
+package storage
+
+import (
+	"crypto"
+	"fmt"
+	vault "github.com/hashicorp/vault/api"
+	"github.com/nuts-foundation/nuts-node/crypto/util"
+)
+
+const privateKeyPath = "nuts-private-keys"
+const kvEnginePath = "kv"
+const keyName = "key"
+
+type logicaler interface {
+	Read(path string) (*vault.Secret, error)
+	Write(path string, data map[string]interface{}) (*vault.Secret, error)
+}
+
+type vaultKVStorage struct {
+	client logicaler
+}
+
+// NewVaultKVStorage creates a new Vault backend using the kv version 1 secret engine: https://www.vaultproject.io/docs/secrets/kv
+// It currently only supports token authentication which should be provided by the token param.
+// If vaultAddr is empty, the VAULT_ADDR environment should be set.
+// If token is empty, the VAULT_TOKEN environment should be is set.
+func NewVaultKVStorage(token string, vaultAddr string) (Storage, error) {
+	client, err := configureVaultClient(token, vaultAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	vaultStorage := vaultKVStorage{client: client.Logical()}
+	err = vaultStorage.checkConnection()
+	return vaultStorage, err
+}
+
+func configureVaultClient(token, vaultAddr string) (*vault.Client, error) {
+	config := vault.DefaultConfig()
+	client, err := vault.NewClient(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to initialize Vault client: %w", err)
+	}
+	client.SetToken(token)
+	if vaultAddr != "" {
+		if err = client.SetAddress(vaultAddr); err != nil {
+			return nil, fmt.Errorf("vault address invalid: %w", err)
+		}
+	}
+	return client, nil
+}
+
+func (v vaultKVStorage) checkConnection() error {
+	secret, err := v.client.Read("auth/token/lookup-self")
+	if err != nil {
+		return fmt.Errorf("unable to connect to Vault: unable to retrieve token status: %w", err)
+	}
+	if len(secret.Data) == 0 {
+		return fmt.Errorf("could not read token information on auth/token/lookup-self")
+	}
+	return nil
+}
+
+func (v vaultKVStorage) GetPrivateKey(kid string) (crypto.Signer, error) {
+	path := fmt.Sprintf("%s/%s/%s", kvEnginePath, privateKeyPath, kid)
```

Comment:
We need to make sure `kid` doesn't contain slashes otherwise someone would be able to read arbitrary data from the Vault key value store.

Reply:
(if it contains `../` for example)

Reply:
Same for `PrivateKeyExists` and `SavePrivateKey`

Reply:
Also: how does Vault handle the key fingerprint after the `#`? Is it ignored or part of the identifier?

Reply:
Tested, the part after the `#` or _fragment_ is not ignored and returns different results. Also, I have added some path sanitation so the “dot-dot-slash” aka “directory traversal” attack is mitigated.